Fix send/recv lost spill block

behlendorf opened the pull request:
### Motivation and Context

Issue #5290 
Issue #6224
OpenZFS-issue: https://www.illumos.org/issues/9952
FreeBSD-issue: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=233277

### Description __[revised]__

When receiving a DRR_OBJECT record the receive_object() function needs to determine how to handle a spill block associated with the object.  It may need to be removed, preserved unmodified, or will updated by a subsequent DRR_SPILL record.
    
This determination is currently accomplished using a heuristic which takes in to account the DRR_OBJECT record and the existing object properties.  This is a problem because there isn't quite enough information available to do the right thing under all circumstances.
    
What's need to resolve this is an additional flag in the DRR_OBJECT which indicates if the object being received references a spill block.  If it does, then a spill block referenced by the existing object should always be preserved.  Either it will need to be updated or will be replaced by a subsequent DRR_SPILL record.  Conversely, if the object being received doesn't reference a spill block then any existing spill block should always be removed.
    
The DRR_SPILL_BLOCK bit was added for this purpose along with the new DMU_BACKUP_FEATURE_SPILL_BLOCK feature.  Unfortunately, since previous versions of ZFS will not understand this feature it cannot be enabled unconditionally without breaking backwards compatibility.  This would be highly disruptive to the existing install base.
    
Therefore, this feature is not enabled by default when creating a non-raw send stream.  Instead, with this change applied additional DRR_SPILL records will be inserted in to the stream.  This has the advantage of being fully backward compatible.  Existing ZFS systems receiving this stream will recreate the spill block if it was incorrectly removed.
    
The downside of this approach is that even though spill blocks are generally small it may increase the size of the stream, and of the received snapshot.  Additionally, when receiving stream from old versions of ZFS spill blocks may still be lost.  For this reason, the DMU_BACKUP_FEATURE_SPILL_BLOCK feature (which has no such downsides) will be enabled in a future release.
    
Finally, raw sends always enable the DMU_BACKUP_FEATURE_SPILL_BLOCK feature.  This is required since the raw encrypted buffer must be written exactly as received or they will result in authentication errors and be inaccessible.  Since raw sends are a new feature in this send release, this should not be a significant problem.
    
### How Has This Been Tested?

The `send_realloc_files` and `send_realloc_encrypted_files` test cases, which generate random incremental streams, were updated to force spill blocks and validate their contents where received.  Prior to making this change the updated test cases would detect a problem within a handful of incremental send/recv cycles.  With the fix applied, and the test cases updated to run 50 send/recv incrementals per-test run, these test cases have so far been running for over 24 hours without detect any problems.

```
Test: tests/functional/rsend/setup (run as root) [00:03] [PASS]
Test: tests/functional/rsend/send_realloc_files (run as root) [28:30] [PASS]
Test: tests/functional/rsend/send_realloc_encrypted_files (run as root) [31:29] [PASS]
Test: tests/functional/rsend/send_spill_block (run as root) [00:24] [PASS]
Test: tests/functional/rsend/cleanup (run as root) [00:00] [PASS]
Test: tests/functional/rsend/setup (run as root) [00:03] [PASS]
Test: tests/functional/rsend/send_realloc_files (run as root) [28:15] [PASS]
Test: tests/zfs-tests/tests/functional/rsend/send_realloc_encrypted_files (run as root) [31:45] [PASS]
Test: tests/functional/rsend/send_spill_block (run as root) [00:24] [PASS]
Test: tests/functional/rsend/cleanup (run as root) [00:00] [PASS]
...
```

An additional `send_spill_block` test case was added specifically for spill blocks.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

lundman commented:
Apologies for the noise - just curious if this is in some way related to the Solaris send FEATURE:

```
#define DMU_BACKUP_FEATURE_SPILLBLOCKS  (0x20)
         This change is to allow important fix in ZFS to avoid metadata
         corruptions related to ACL.  An upgrade to same or greater version
         of ZFS is required in destination for ZFS streams to work properly.
```
since OpenZFS never got this afaik, and yet there must have been a need for it at some point?

behlendorf commented:
@lundman no apology necessary.  Yes, this sounds like exactly the same issue since on Solaris ACLs would be stored in the spill block.

Based on the description it sounds like they decided to address this on the receive side.  @tcaputi and I looked in to a similar solution and came to the same conclusion that a new `DMU_BACKUP*` flag would be required.  For compatibility reasons this was something I wanted to avoid if possible.  Unfortunately, we can only guess at exactly how they use this new flag in the code.

matveevandrey commented:
Will this fix be included to 0.7.14?

behlendorf commented:
> Will this fix be included into 0.7.14?

Yes, I think we should look in to backporting a version of the finalized change for 0.7.14.

tcaputi commented on module/zfs/dmu_send.c:
```diff
@@ -587,15 +597,54 @@ dump_dnode(dmu_sendarg_t *dsp, const blkptr_t *bp, uint64_t object,
+	 *
+	 * The feature is mandatory for raw sends to prevent authentication
+	 * errors in dnode blocks.  It is optional for non-raw sends and
+	 * disabled by default.  It will be enabled by default in a future
```
nit: disabled by default for backwards compatibility.

tcaputi commented on module/zfs/dmu_send.c:
```diff
@@ -587,15 +597,54 @@ dump_dnode(dmu_sendarg_t *dsp, const blkptr_t *bp, uint64_t object,
+	 */
+	if (dsp->dsa_featureflags & DMU_BACKUP_FEATURE_SPILL_BLOCK) {
+		if (dnp->dn_flags & DNODE_FLAG_SPILL_BLKPTR)
+			drro->drr_flags |= DRR_SPILL_BLOCK;
```
nit: this could be combined to a single `if` statement.

ahrens commented:
I have a few high level questions after reading the description.  Apologies if these will become clear once I read the diff...

> When receiving a DRR_OBJECT record the receive_object() function needs to determine how to handle a spill block associated with the object. It may need to be removed, preserved unmodified, or will updated by a subsequent DRR_SPILL record.

In the case where it will be updated by a subsequent SPILL record, does it matter whether the OBJECT record causes us to remove vs preserve the SPILL block?  (I'm hoping not.)

In the case where the sending system doesn't have a spill block, and the receiver should remove the spill block, what is the negative impact of not removing it?  I'm hoping it's negligible, in which case I wonder if we could always try to preserve the spill block (when we aren't absolutely sure it is not needed).

>  This is a problem because there isn't quite enough information available to do the right thing under all circumstances.

Do you know what specifically will leads to incorrect behavior?

> The downside of this approach is that even though spill blocks are generally small it may increase the size of the stream, and of the received snapshot. 

I think that's a very tiny downside, considering that we're already sending the OBJECT record and updating the dnode.

> when receiving stream from old versions of ZFS spill blocks may still be lost

Does this change make it any worse than it already is?

> the DMU_BACKUP_FEATURE_SPILL_BLOCK feature (which has no such downsides) will be enabled in a future release

I think you mean `enabled *by default*`?  I think that would be the first time that we make zfs send backwards-incompatible (can't be received by older software) by default.  If this was controlled by a CLI flag to `zfs send`, that might be OK.  In my opinion, having to change a module parameter to do any send to older software is not OK.

Would it be possible to include the extra information (about presence/absence of the spill block) in the send stream in a way that new software can take advantage of it, but old software will ignore it (and have the same problems it does today)?

amotin commented:
I'll look deeper, but on a first quick look it seems good to me.  The only thing that scratches my eye is use of tunable to control this feature. `zfs send` already has several flags controlling replication stream features, that could be set depending on receiving side version, and use of different and global mechanism here looks bad to me.

behlendorf commented:
> In the case where it will be updated by a subsequent SPILL record, does it matter whether the OBJECT record causes us to remove vs preserve the SPILL block? (I'm hoping not.)

As long as we know it's going to be updated by a SPILL record you're right it doesn't matter what we do in `receive_object()`.  I believe the code makes this more clear.

> In the case where the sending system doesn't have a spill block, and the receiver should remove the spill block, what is the negative impact of not removing it? I'm hoping it's negligible, in which case I wonder if we could always try to preserve the spill block (when we aren't absolutely sure it is not needed).

For a non-raw send stream the consequences have so far been minimal.  You may end up with a stale spill block, but in the testing I've done that hasn't led to any real problems.   All the expected SAs can be found in the bonus buffer, so that stale spill block never gets accessed.  Let me find out if there are other side effects from never removing it when reallocate case.

The stories a little different for raw send stream since any discrepancy is correctly detected and reported as an authentication error.  This makes all the objects in that dnode block inaccessible.  This is particularly bad, when it's `blkid=0` since you're then unable to mount the dataset.  For this reason, the new flag does need to be mandatory when using raw sends.

> Do you know what specifically will leads to incorrect behavior?

The most likely case is the one described in the original issue, https://www.illumos.org/issues/9952, where only the block size changes.  In which case the object may have been freed/allocated or merely appended too, we don't have enough information to know.

> I think that's a very tiny downside, considering that we're already sending the OBJECT record and updating the dnode.

That's my feeling as well.  Plus it has the advantage of fixing the issue on the send side which means even receiving systems which aren't updated won't loss their spill blocks.

> Does this change make it any worse than it already is?

It does not, when receiving a stream from a system without this change the behavior is unchanged.

> In my opinion, having to change a module parameter to do any send to older software is not OK.

This was exactly my concern.  For this reason, I opted to make the default behavior for non-raw send streams to include the extra spill blocks in the stream.  This doesn't break backwards compatibility with any existing systems, and forces the legacy systems to do a reasonable thing (not lose spill blocks).  But as mentioned above it doesn't allow the updated system to do any better job when it's receiving a stream from a non-updated system. 

> Would it be possible to include the extra information (about presence/absence of the spill block) in the send stream in a way that new software can take advantage of it, but old software will ignore it (and have the same problems it does today)?

We definitely could.  This change is only incompatible because I used the normal mechanism of adding a new `DMU_BACKUP_FEATURE_*` flag to indicate that the bit describing the presence/absence of the spill block is valid.  We could move this information to the stream header nvlist.  Then older versions of ZFS would ignore the new field and be no worse off, newer versions could detect the functionality and do the right thing.  Would this be preferable?  My only reluctance is it that behavior would be different from all the other stream features (personally, I do think it would be worth it).

> I'll look deeper, but on a first quick look it seems good to me. The only thing that scratches my eye is use of tunable to control this feature.

I can't say I care for it myself.  What if instead I changed the implementation as described above so this feature was always enabled (and ignored on older releases).  Then the module option could be removed, or changed to a toggle to enable/disable sending extra spill blocks.  In which case would we want to disable it by default?

amotin commented:
> I can't say I care for it myself. What if instead I changed the implementation as described above so this feature was always enabled (and ignored on older releases). Then the module option could be removed, or changed to a toggle to enable/disable sending extra spill blocks. In which case would we want to disable it by default?

Having backward compatible stream would be great, if that is possible for present (and past) receive code on all platforms.  What's about tunable for sending extra spill blocks, it is not as bad for me as original tunable controlling stream compatibility, since I don't think many people would care about it.  I suppose the only reason to disable it is to save some stream size in case of sending many dnodes with the spill blocks, but it does not sound very dramatic to me. May be we could keep it enabled for few years, then disable and in few more years remove completely.

behlendorf commented:
Based on the feedback above I've refreshed this PR and made the following significant changes.

* The `DMU_BACKUP_FEATURE_SPILL_BLOCK` feature flag was converted to a `DRR_FLAG_SPILL_BLOCK` flag which is set in the `drr_begin->drr_flags` fields.  This has the advantage of always keeping the stream backwards compatible.  Previous versions of ZFS will ignore the `DRR_FLAG_SPILL_BLOCK` flag and the `DRR_SPILL_BLOCK` bit in the records.  Updated versions with this change will correctly interpret the new bits and do the right thing.

* When including additional spill blocks in the stream the extra `DRR_SPILL` records set `DRR_SPILL_BLOCK` in their `drr_flags`.  This allows for updated systems to identify and discard the extra spill blocks when they are not required.  Older versions of ZFS will process these records.  This means that adding the extra blocks always results in the optimal behavior when receiving and is therefore enabled by default.  This behavior can still be disabled by setting `zfs_send_extra_spill_blocks=0` if for some reason the need arises.

* Added a new `zfs recv` error message which is printed when attempting to receive a raw stream without the `DRR_FLAG_SPILL_BLOCK` flag set.  This will only effect users of the 0.8.0-rc releases using raw sends.

* The core logic in `receive_object()` remains unchanged when interpreting the new `DRR_SPILL_BLOCK`.  Thus far I've observed no failures after 100 hours of continual incremental send/recvs using the previous patch.

* Documentation updated (man page, comments, etc).

ahrens commented on module/zfs/dmu_recv.c:
```diff
@@ -702,6 +714,9 @@ dmu_recv_resume_begin_sync(void *arg, dmu_tx_t *tx)
 	}
 
+	if (drrb->drr_flags & DRR_FLAG_SPILL_BLOCK)
+		drba->drba_cookie->drc_spill = B_TRUE;
```
this won't be needed once moved to dmu_recv_begin()

ahrens commented on module/zfs/dmu_recv.c:
```diff
@@ -1321,11 +1346,19 @@ receive_object(struct receive_writer_arg *rwa, struct drr_object *drro,
 		    drro->drr_bonustype, drro->drr_bonuslen,
-		    dn_slots << DNODE_SHIFT, tx);
+		    dn_slots << DNODE_SHIFT,
+		    rwa->spill ? DRR_HAS_SPILL_BLOCK(drro->drr_flags) : 0, tx);
```
0 -> B_FALSE

ahrens commented on module/zfs/dmu_send.c:
```diff
@@ -1084,10 +1131,12 @@ dmu_send_impl(void *tag, dsl_pool_t *dp, dsl_dataset_t *to_ds,
 	dsp->dsa_featureflags = featureflags;
 	dsp->dsa_resume_object = resumeobj;
 	dsp->dsa_resume_offset = resumeoff;
+	dsp->dsa_spill = zfs_send_set_spill_block_bit;
```
why bother with this, vs. using zfs_send_set_spill_block_bit in place of dsa_spill?

behlendorf replied:
I wanted to ensure this value would not change while creating the send stream.  Since `zfs_send_set_spill_block_bit` is a global and it could be (but currently isn't) available as a module option.  I primarily added this flag to make my testing easier.  Since I can't think of a reason we'd want to disable this, I'll remove it and `zfs_send_set_spill_block_bit`.

ahrens commented on module/zfs/dmu_send.c:
```diff
@@ -64,6 +64,10 @@ int zfs_send_corrupt_data = B_FALSE;
 /* Set this tunable to FALSE to disable setting of DRR_FLAG_FREERECORDS */
 int zfs_send_set_freerecords_bit = B_TRUE;
+/* Set this tunable to FALSE to disable setting of DRR_FLAG_SPILL_BLOCK */
+int zfs_send_set_spill_block_bit = B_TRUE;
```
There isn't a module_param for this.  Is that intentional?

ahrens commented on module/zfs/dmu_send.c:
```diff
@@ -587,15 +597,49 @@ dump_dnode(dmu_sendarg_t *dsp, const blkptr_t *bp, uint64_t object,
+	/*
+	 * Send extra DRR_SPILL records.  By default, these are included in
+	 * the stream in order to recreate spill blocks that are incorrectly
+	 * removed.  This may occur in previous versions of ZFS when certain
```
What is the difference in correctness (if any) between receiving a non-raw stream (generated by the new code with the default tunables) on old software vs on new software (with this change)?

behlendorf replied:
With this change both old and new software receiving the non-raw stream will not have any differences in correctness.  The one caveat to this, is that because we're always sending the spill block additional versions of the exact same spill block will be created on the old software.  The new software will correctly preserve existing spill blocks and discard the extra DRR_SPILL records.

behlendorf commented on include/sys/zfs_ioctl.h:
```diff
@@ -160,16 +160,35 @@ typedef enum dmu_send_resume_token_version {
+/*
+ * When DRR_FLAG_SPILL_BLOCK is set it indicates the DRR_SPILL_BLOCK flag
+ * is valid in the send stream.  A single bit was shared for this purpose
+ * because drr_begin->drr_flags is only a uint8_t.
```
> drr_begin->drr_flags is uint32_t (drr_object->drr_flags is uint8_t)

I can try and make this comment clearer.  There is plenty of room in `drr_begin->drr_flags`.  My concern was with `drr_object->drr_flags` and `drr_spill->flags` which are both a uint8_t.  See next comment.

> I don't see where this bit-sharing is taking place?

The sharing referred to is that `DRR_SPILL_BLOCK` can be set for both `drr_object` and `drr_spill` records.  But it has a different meaning for each record type.  Are we OK with this, or should we use another bit and define a `DRR_HAS_SPILL_BLOCK` for drr_object flags and a `DRR_EXTRA_SPILL_BLOCK` for drr_spill flags.

> Is the past tense ("was") relevant here?

No, I'll update it.

pcd1193182 replied:
Personally I'm ok with the current scheme(using the same bit in each type, with different meanings), but the language in this comment does need some clarification. I'll provide my rough draft below, you can pick and choose bits you like, or ignore it if you prefer:

```
When DRR_FLAG_SPILL_BLOCK is set it indicates the DRR_SPILL_BLOCK flag is meaningful in the send stream.

When DRR_FLAG_SPILL_BLOCK is set, DRR_OBJECT records will have DRR_SPILL_BLOCK set if and only if they should have a spill block (either an existing one, or a new one in the send stream).  When clear the object does not have a spill block and any existing spill block should be freed.

Similarly, when DRR_FLAG_SPILL_BLOCK is set, DRR_SPILL records will have DRR_SPILL_BLOCK set if and only if they were included for backward compatibility purposes, and can be safely ignored by new version of zfs receive.  Previous versions of ZFS which do not understand the DRR_FLAG_SPILL_BLOCK will process this record and recreate any missing spill blocks.

DRR_SPILL_BLOCK has different meanings in different record types because the meanings are closely linked, there is per-record-type specific behavior, and there is only a small amount of space (8 bits) in drr_flags in the drr_object and drr_spill structures.
```

behlendorf replied:
Thanks, I opted to use your updated comment.

behlendorf commented:
Updated to address the review feedback.  One additional change included was renaming `zfs_send_extra_spill_blocks` to `zfs_send_unmodified_spill_blocks`.

behlendorf commented on module/zfs/dmu_send.c:
```diff
@@ -587,15 +597,49 @@ dump_dnode(dmu_sendarg_t *dsp, const blkptr_t *bp, uint64_t object,
+	 */
+	if (zfs_send_extra_spill_blocks &&
+	    (dnp->dn_flags & DNODE_FLAG_SPILL_BLKPTR) &&
+	    (DN_SPILL_BLKPTR(dnp)->blk_birth <= dsp->dsa_fromtxg)) {
```
There are two main reasons I ended up adding it here:

1) The traversal logic correctly skips blocks born before the specified `td->td_min_txg`.  Rather than try and work an exception in to that code for these unmodified blocks with birth txgs below `td->td_min_txg` it seemed simpler to add it here.
2) We need to ensure the `DRR_SPILL` record occurs after the `DRR_OBJECT` record in the resulting stream.  Again, this made it easy.

If you can suggest a nice way to move this step in to the traversal logic I'm all for it.

pcd1193182 replied:
Yeah, on further inspection, doing it in dmu_traverse would be complex. The better alternative is probably to do it in send_cb (so it goes through the entire send pipeline), but we don't have any of the logic in redacted send integrated yet, which breaks down send_cb calls by type and creates records for each one. Under that model, it would be easy to just create two records when the object record happens.  I'll probably move this into send_cb when I rebase redacted send on top of this, in that case, so don't worry about this comment.

ahrens commented on include/sys/zfs_ioctl.h:
```diff
@@ -160,16 +160,41 @@ typedef enum dmu_send_resume_token_version {
  */
 #define	DRR_CHECKSUM_DEDUP	(1<<0) /* not used for DRR_SPILL blocks */
 #define	DRR_RAW_BYTESWAP	(1<<1)
+#define	DRR_SPILL_BLOCK		(1<<2)
```
The name of this macro is not that descriptive.  What would you think about adding 2 macros, one for each use:
```
DRR_OBJECT_HAS_SPILL 1<<2 // this OBJECT record has a spill block
DRR_SPILL_IS_UNMODIFIED 1<<2  // this SPILL record describes a block that is not actually modified, and can be ignored by software that proceses DRR_OBJECT_HAS_SPILL
```

behlendorf replied:
Sounds good, I opted to name the flags `DRR_OBJECT_SPILL` and `DRR_SPILL_UNMODIFIED`, and check macros `DRR_OBJECT_HAS_SPILL()` and `DRR_SPILL_IS_UNMODIFIED`.  I think this strikes a good balance between descriptive names while remaining  consistent with the existing naming scheme.  For example, only using `IS` or `HAS` in the macros which check the flag.

ahrens commented on include/sys/zfs_ioctl.h:
```diff
@@ -160,16 +160,41 @@ typedef enum dmu_send_resume_token_version {
 
 #define	DRR_IS_DEDUP_CAPABLE(flags)	((flags) & DRR_CHECKSUM_DEDUP)
 #define	DRR_IS_RAW_BYTESWAPPED(flags)	((flags) & DRR_RAW_BYTESWAP)
+#define	DRR_HAS_SPILL_BLOCK(flags)	((flags) & DRR_SPILL_BLOCK)
```
if we make the above change, probably want to do something similar here, e.g.:
```
DRR_OBJECT_HAS_SPILL_FLAG(flags)
DRR_SPILL_IS_UNMODIFIED_FLAG(flags)
```

behlendorf commented:
> Could you please do some manual tests sending and receiving between old and new software, to verify that everything works as expected? e.g:

Absolutely, in fact this is something I should have mentioning I did test in the original PR.  The testing was done by disabling the relevant sections of this change to effectively use old or new code as required by the test.

> old software to new software (non-raw), receives OK although bug may still be present

Verified by disabling changes in `dmu_send.c`.  Test case `send_realloc_files` fails after a handful of iterations (as expected)

> old software to new software raw send, receive errors out

Verified by disabling changes in `dmu_send.c`.  Test case `send_realloc_encrypted_files` failed with expected new error message from this PR.

> new software to old software (non-raw), bug is fixed due to "extra" SPILL records

Verified by disabling changes in `dmu_recv.c`.  Test case `send_realloc_files` has run for multiple hours without failure.

> new software to new software, raw and non-raw sends.

 Test cases `send_realloc_files` and `send_realloc_encrypted_file" both run for multiple hours without failures.
